[BOLT][AArch64] Add call relaxation pass

maksfb opened the pull request:
Add a function call relaxation pass that groups functions into clusters, each with a maximum size of 128MB. This pass is enabled using the `--relax-exp` option.

* Within each cluster: function calls do not require relaxation.

* Between clusters: thunks are created at the cluster boundaries to handle inter-cluster calls. These thunks can be either short or long, depending on the proximity of the thunk destination. When possible, a long thunk may be shared by both adjacent clusters.

* Hot functions: if all hot functions (after reordering) fit within a single cluster (under 128MB), no thunks are needed on the hot path, as determined by the profile.

* PLT behavior: for the Procedure Linkage Table (PLT), it is currently assumed that the hottest cluster will be placed adjacent to the PLT, so thunks are not required for that cluster. This assumption can be disabled with `--relax-plt=0` option.

* Pass behavior: when enabled, this pass replaces the LongJmp pass. It typically runs much faster because it does not attempt to predict the exact code layout.

* Local function branches are relaxed using the compact code model (`--compact-code-model`).

maksfb committed:
[BOLT][AArch64] Add experimental call relaxation pass

Add a function call relaxation pass that groups functions into clusters,
each with a maximum size of 128MB. This pass is enabled using the
--relax-exp` option.

* Within each cluster: function calls do not require relaxation.

* Between clusters: thunks are created at the cluster boundaries to
handle inter-cluster calls. These thunks can be either short or long,
depending on the proximity of the thunk destination. When possible,
a long thunk may be shared by both adjacent clusters.

* Hot functions: if all hot functions (after reordering) fit within a
single cluster (under 128MB), no thunks are needed on the hot path, as
determined by the profile.

* PLT behavior: for the Procedure Linkage Table (PLT), it is currently
assumed that the hottest cluster will be placed adjacent to the PLT, so
thunks are not required for that cluster. This assumption can be
disabled with `--relax-plt=0` option.

* Pass behavior: when enabled, this pass replaces the LongJmp pass. It
typically runs much faster because it does not attempt to predict the
exact code layout.

paschalis-mpeis commented on bolt/lib/Passes/LongJmp.cpp:
```diff
@@ -944,13 +953,307 @@ void LongJmpPass::relaxLocalBranches(BinaryFunction &BF) {
+  // Conservatively estimate emitted function size. Assume the worst case
+  // alignment.
+  auto estimateFunctionSize = [&](const BinaryFunction &BF) -> uint64_t {
+    // Conservative estimation of the aligned function size.
```
nitpick: could drop, explained above?

```suggestion
```

paschalis-mpeis commented on bolt/lib/Passes/LongJmp.cpp:
```diff
@@ -944,13 +953,307 @@ void LongJmpPass::relaxLocalBranches(BinaryFunction &BF) {
+        auto It = PLTCluster.Callees.find(BF.getSymbol());
+        if (It != PLTCluster.Callees.end())
+          PLTCluster.Callees.erase(It);
+      }
```
Does this mean that some PLT function calls another PLT function?

Do we expect such cases or shall we convert this to an assertion instead?

maksfb replied:
Here we remove the PLT function from the callee list of the cluster. I.e., we are pretending that PLT is part of this cluster such that thunks will not be generated when not necessary.

paschalis-mpeis commented on bolt/lib/Passes/LongJmp.cpp:
```diff
@@ -944,13 +953,307 @@ void LongJmpPass::relaxLocalBranches(BinaryFunction &BF) {
+    }
+  }
+
+  /// Create a thunk with +-128MB span.
```
nit: here and on line [1106](https://github.com/llvm/llvm-project/pull/173952/files#diff-937cd38690384c2c6f0b6b4856851df783b2994e4bd5622581cae96020072c78R1106)
```suggestion
  // Create a thunk with +-128MB span.
```

paschalis-mpeis commented on bolt/lib/Passes/LongJmp.cpp:
```diff
@@ -944,13 +953,307 @@ void LongJmpPass::relaxLocalBranches(BinaryFunction &BF) {
+        ThunkCluster = &Clusters[ClusterNum - 1];
+      ThunkCluster->ThunkList.push_back(Thunk);
+
+      // Register thunks for all symbols associated with the function.
```
Checking my understanding here:
- the if case goes for multiple entry points, when it's not the first entry point
- the else case handles either a single entry point (where there is only one symbol anyway), or the first entry point of a multiple entry-point case?

paschalis-mpeis commented on bolt/lib/Passes/LongJmp.cpp:
```diff
@@ -944,13 +953,307 @@ void LongJmpPass::relaxLocalBranches(BinaryFunction &BF) {
+  for (FunctionCluster &FC : Clusters) {
+    for (BinaryFunction *BF : FC.Functions) {
+      if (!CallMap.count(BF))
+        continue;
```
TMU, we still iterate over instructions of functions that don't do cross-cluster calls.

Maybe in a future patch we could skip those cases as well, by checking `FC.Callees`?

paschalis-mpeis commented on bolt/test/AArch64/relax-exp.s:
```diff
@@ -0,0 +1,78 @@
+## Additionally, check veneers: no double veneers in lite mode and proper names
+## for BOLT-introduced veneers.
+
+# RUN: llvm-mc -filetype=obj -triple aarch64-unknown-unknown %s -o %t.o
```
Can we also check on the number of clusters build per method? ie:

```
built N function cluster(s)
```

paschalis-mpeis reviewed (commented):
Thanks for the patch Maksim. Great work! 🙌 

I left some comments and a few questions below.

#### Clusters with lite-mode off
With lite-mode enabled, foo seems to be using three long thunks to reach callees, which makes sense.

With lite-mode disabled, I might be missing something.
We end up with 3 clusters for the .cold section, which also seems reasonable: one for each of foo, bar, and _start.

In the .text section, `hot` calls `foo` directly.
That looks OK, since it happens to be within range (.cold comes right after the hot section).

What I'm less sure about is that hot appears to be using the ThunkList from the first cold cluster? ie a short thunk to reach `bar` and a long one to reach `_start`.

Minimal example:

```assembly
.text
<hot>:
  ..
  bl <__AArch64Thunk_bar>
  .. ; code after call-to-thunk

.text.cold:

<foo>:
	.. ; code ~112MB

// ThunkList that contains foo
<__AArch64Thunk_bar>:
  b	<bar> ; short thunk to reach bar

...

```

Could .text run into issues if the code after the call-to-hunk grows, such that the distance to the ThunkList exceeds 128MB? A larger .text section or a larger first cold cluster could cause this.

Also, was lite=0 tested on your workloads?


#### Discussion on planning

Based on last Office Hours, this has been tested at Meta and already covers cases that LongJmp cannot.

The plan would be to introduce these experimental flags and keep it that way for some time, so community can try both flags (function/plt clustering) before making it the default?

Eventually, the existing LongJmp could become a legacy alternative, with compact-code and relax-exp flags folded into the new pass?

We don't have many binaries of this size ourselves, so if feasible it might be helpful for users with such examples to list them somehow, eg:
ie Company had 2 big binaries:
1. XMiB, N Clusters, with S short and L long thunks
2. YMiB, ...

maksfb committed:
fixup! [BOLT][AArch64] Add experimental call relaxation pass

maksfb commented:
Thank you for a thorough review, Paschalis! I've addressed some of your comments and more are in the pipeline.

> We don't have many binaries of this size ourselves, so if feasible it might be helpful for users with such examples to list them somehow

Regarding our internal binaries, almost all of them are over 128MB of code, i.e. it would be easier to list the ones that are less than that. One such binary would be Clang :) For the lite mode however, we are able to squeeze all hot code into just one cluster so far.

> Also, was lite=0 tested on your workloads?

We build HHVM without lite mode. Stats with an older version of the patch:
```
Built 2 clusters
  Cluster: 0
    269188 functions
    13220 callees
    121973928 bytes
  Cluster: 1
    217129 functions
    39580 callees
    131064744 bytes
BOLT-INFO: 33613 thunks created

yozhu commented on bolt/lib/Passes/LongJmp.cpp:
```diff
@@ -944,13 +953,306 @@ void LongJmpPass::relaxLocalBranches(BinaryFunction &BF) {
+    if (!BC.shouldEmit(BF))
+      return 0;
+    uint64_t Size = BF.estimateSize();
+    if (BF.hasValidIndex())
```
We will count a function's code size regardless of whether it has valid index. Should we also skip the check when adding worse case padding due to alignment?

maksfb replied:
The hidden assumption here is that a cold function (i.e. without a valid index) does not need an alignment. I will refactor the code to eliminate this assumption.

maksfb replied:
I've addressed the issue in 4b3d32f2d670b776f042be78c20103f9558663ee

yozhu commented on bolt/lib/Passes/LongJmp.cpp:
```diff
@@ -944,13 +953,306 @@ void LongJmpPass::relaxLocalBranches(BinaryFunction &BF) {
+    }
+  }
+
+  if (NumShortThunks)
```
Just a small suggestion - could we always output number of short (and long) thunks created, even if it is zero?

maksfb committed:
[BOLT] Improve code size estimation for call relaxation pass

maksfb commented:
> What I'm less sure about is that hot appears to be using the ThunkList from the first cold cluster? ie a short thunk to reach bar and a long one to reach _start.
> ...
> Could .text run into issues if the code after the call-to-hunk grows, such that the distance to the ThunkList exceeds 128MB? A larger .text section or a larger first cold cluster could cause this.

`hot` will use thunks from the cluster shared with `foo` (which you call cold) only as long as `hot` and `foo` get allocated in the same cluster. On the other hand, if the code after the call-to-thunk grows over ~13MB, `foo` will get allocated into a new cluster, while the first cluster will be cut short, meaning `hot` will use thunks that are allocated before `foo`.

rafaelauler committed:
Merge branch 'main' into gh-relax-exp-pass

rafaelauler committed:
Fix botched merge attempt + format

labrinea commented:
SGTM, feel free Rafael.

labrinea commented:
> SGTM, feel free Rafael.
> 
> Just noting that I am getting fix up values out of range and jit link errors using this patch on chromium trained with speedometer. I haven't got to the bottom of it yet. I believe it is due to calls from bolt emitted text back to the original text. I played with the relax-plt option and lite mode in my experiments.
> 
> **lite-mode = 1, relax-plt = 1** (default)
> 
> ```
> BOLT-INFO: built 1 function cluster(s)
> BOLT-INFO: cluster: 0
> BOLT-INFO:   38397 function(s)
> BOLT-INFO:   11438 callee(s)
> BOLT-INFO:   24845196 estimated bytes
> BOLT-INFO: 11048 long thunks created
> BOLT-ERROR: JITLink failed: In graph in-memory object file, section .text: relocation target 0x102f70a0 (ResolveRemaskPointer/1@PLT) is out of range of Branch26PCRel fixup at address 0x102f70a0 (_ZN5media8TuneableIiED2Ev/1, 0x1cc00000 + 0x620)
> ```
> 
> **lite-mode = 1, relax-plt = 0**
> 
> ```
> BOLT-INFO: built 1 function cluster(s)
> BOLT-INFO: cluster: 0
> BOLT-INFO:   38397 function(s)
> BOLT-INFO:   11438 callee(s)
> BOLT-INFO:   24845196 estimated bytes
> BOLT-INFO: 11438 long thunks created
> BOLT-ERROR: JITLink failed: In graph in-memory object file, section .text: relocation target 0x8ba7048 (_ZN6icu_7717CollationIterator17getCE32FromPrefixEPKNS_13CollationDataEjR10UErrorCode/1) is out of range of Branch26PCRel fixup at address 0x8ba7048 (_ZN5media8TuneableIiED2Ev/1, 0x1cc00000 + 0x43bcc)
> ```
> 
> **lite-mode = 0, relax-plt = 0**
> 
> ```
> BOLT-INFO: built 2 function cluster(s)
> BOLT-INFO: cluster: 0
> BOLT-INFO:   246898 function(s)
> BOLT-INFO:   13947 callee(s)
> BOLT-INFO:   125824716 estimated bytes
> BOLT-INFO: cluster: 1
> BOLT-INFO:   212397 function(s)
> BOLT-INFO:   23001 callee(s)
> BOLT-INFO:   98055112 estimated bytes
> BOLT-INFO: 32213 short thunks created
> BOLT-INFO: 4035 long thunks created
> BOLT-ERROR: JITLink failed: In graph in-memory object file, section .text.cold: relocation target 0x291fcbc (FUNCat0x291fcbc) is out of range of Branch26PCRel fixup at address 0x291fcbc (__aarch64_ldadd4_acq_rel.cold.0/1, 0x1d991ec0 + 0x94f7ac)
> ```
> 
> The original text is about 220MB, so if I understand the assumptions of the pass correctly `lite mode = 0` is expected to fail (if hot text > 128MB), so I am mostly worried about the other cases.

Interestingly I found that JITLink maps both R_AARCH64_CALL26 and R_AARCH64_JUMP26 to the same edge kind: Branch26PCRel but relaxCalls() only collects instructions where BC.MIB->isCall(Inst) is true. I replaced getFunctionForSymbol with this lambda and it succeeded:

```
auto getDirectFunctionTarget = [&](const MCInst &Inst) -> const MCSymbol * {
  if (BC.MIB->isCall(Inst)) {
    if (BC.MIB->isIndirectCall(Inst) || BC.MIB->isIndirectBranch(Inst))
      return nullptr;
  } else if (BC.MIB->isUnconditionalBranch(Inst)) {
    if (BC.MIB->isIndirectBranch(Inst))
      return nullptr;
  } else {
    return nullptr;
  }  
  const MCSymbol *TargetSymbol = BC.MIB->getTargetSymbol(Inst);
  if (!TargetSymbol)
    return nullptr;
  
  if (!BC.getFunctionForSymbol(TargetSymbol))
    return nullptr;
 
  return TargetSymbol;
};
```
